agent/Shooter.py

```diff
@@ -85,23 +104,29 @@ def select_card(self,
         else:
             # We are following a trick
             if self.is_early(partial_state):
-                # TODO is_early most likely counts too much time when it comes to whether
-                # we should follow with low cards or not so we can update it later
                 lowest = self.get_lowest(self.cards_in_hand)
                 return random.choice(lowest)
             else:
                 # Later in the game so try to take tricks from players
-                # TODO we should be checking if we can even follow suit to know
                 # if we are capable of taking the trick
                 if (self.is_last):
                     # We know the trick is over after this action so do the bare minimum to win
-                    lowest_high = self.lowest_high(partial_state, self.cards_in_hand)
-                    return random.choice(lowest_high)
+                    if self.following_lead(partial_state, self.cards_in_hand):
+                        # We can follow suit so take it
+                        lowest_high = self.lowest_high(partial_state, self.cards_in_hand)
+                        return random.choice(lowest_high)
+                    else:
+                        # We could not follow suit so do not bother trying to win
+                        lowest = self.get_lowest(self.cards_in_hand)
                 else:
                     # Not last so be safe and play highest
-                    highest = self.get_highest(self.cards_in_hand)
-                    return random.choice(highest)
-            return random.choice(self.cards_in_hand)
+                    if self.following_lead(partial_state, self.cards_in_hand):
+                        # We could possibly take it so try to
+                        highest = self.get_highest(self.cards_in_hand)
+                        return random.choice(highest)
+                    else:
+                        # We can not follow so there is no chance of taking it
+                        lowest = self.get_lowest(self.cards_in_hand)
```

Comment:
This never gets used if it gets here. It won't get here until self.following_lead() returns False (which it doesn't right now).
Add a return statement here. (Same as above)